Add an off course vehicle state

arkadyan opened the pull request:
Asana ticket: [Vehicle states: invalid, vehicle is off course](https://app.asana.com/0/1112935048846093/1128793574082090)

![Screen Shot 2019-07-10 at 13 05 10](https://user-images.githubusercontent.com/42339/60989713-a4bc7b80-a314-11e9-8865-dcdd7287ede9.png)

(MTSTY is where both of these are scheduled to be at this moment)

ingridpierre commented:
![image](https://user-images.githubusercontent.com/5806643/61075406-ecf89e00-a3e7-11e9-8c3e-c5bf204fb26f.png)
@arkadyan with this ticket, are we able to add some of the other elements to the vehicle properties panel? We definitely want to update how we're showing adherence when the vehicle isn't doing what it's supposed to.

(ignore the janky exclamation icon, would be happy just to have the text  say "Not available" and "Invalid (Off-Course)

arkadyan commented on assets/src/components/ladder.tsx:
```diff
@@ -128,21 +132,35 @@ const Ladder = ({
+          ladderVehicle.vehicle.scheduledLocation.timepointStatus,
+          ladderVehicle.vehicleDirection
+        )
+      : ladderVehicle.y
```
Yeah, this is good. I put all the properties we need into LadderVehicle that we need so that we could remove the vehicle. I also moved the `y` code all to one place. The component code is much simpler now which I think is a good sign.

arkadyan commented on assets/src/components/ladder.tsx:
```diff
@@ -128,21 +132,35 @@ const Ladder = ({
+        className={`m-ladder__vehicle ${status(
+          ladderVehicle
+        )} ${selectedClass}`}
+        transform={`translate(${ladderVehicle.x},${y})`}
```
Addressed with above.

arkadyan commented:
@skyqrose I addressed your suggestions if you want to take a look now, but I'm also going to make a few small changes to the vehicle properties panel that Ingrid pointed out if you want to wait till those are in.

arkadyan commented:
@ingridpierre Updated the vehicle properties panel:

![Screen Shot 2019-07-12 at 11 13 48](https://user-images.githubusercontent.com/42339/61139179-da896d80-a496-11e9-90ae-39a119a76735.png)

skyqrose commented on assets/css/_vehicle_properties_panel.scss:
```diff
@@ -66,6 +69,10 @@
   }
+
+  .off-course & {
+    color: $color-vehicle-early;
```
I get that this is a red warning color, but it's not really "early". Should we add another color to the ui kit for using in warning cases like this, even if it's the same red as the early color, so we can be more semantic about it here?

(Same for below in `.m-vehicle-properties-panel__not-available {}`)

arkadyan commented on assets/src/components/ladder.tsx:
```diff
@@ -240,41 +240,35 @@ const orientationMatchingVehicle = (
-      />
-    )
-  } else {
+  if (!scheduledY || isUnassignedBySwiftly) {
```
It's needed because scheduledY is always set as long as we have a schedule (which is what the `!scheduledY` case is checking for). So this condition is combining two cases, if there is no schedule (in which case we can't draw a line), or if the trip is unassigned by Swiftly (in which case we are choosing not to draw a line).

skyqrose replied:
For off course vehicles, `y` is already the scheduled location. Can we not set `scheduledY` for off course vehicles?

arkadyan replied:
We could, but again I think that is conflating things in a bad way. "`sheduledY` is null not because we don't have it but because we removed it because that signifies a case where we don't want to draw a line" does not feel like a good way of representing this data to me.

skyqrose replied:
The viewmodel isn't supposed to represent the data, it's supposed to represent what we draw on the screen. (Renaming `y` -> `triangleY` and `scheduledY` -> `scheduledLineY` might make that more clear.)

arkadyan replied:
Yeah I hear what you are saying but I think that that is what it is doing and I like the current names.

arkadyan commented on assets/src/skate.d.ts:
```diff
@@ -79,6 +79,8 @@ export interface Vehicle {
 
 export type VehicleId = string
 
+export type VehicleAdherenceStatus = ScheduleAdherenceStatus | "off-course"
```
My initial reaction is that this feels like conflating two things that aren't really the same thing (we are just using in many of the same places at the moment). I think about headway-based routes for example. Schedule adherence will not be used when we are drawing those, but we'll still want to draw a black triangle for invalid data.

skyqrose replied:
Mm. I see.

Even if the headway view will draw invalid and on-time statuses differently, I think they might still belong together in the data. The statuses don't overlap: a vehicle is exactly one of `on-time`, `late`, `early` or `off-course`, so putting them in the same field/type is a better description of the vehicle and reduces illegal states. The headway view code would have to have an `if (status === "off-course") {black triangle} else {grey triangle}`, but I think that's a fine way for view logic to look at the data, and we basically have that already where the `ladderVehicle` decides whether to draw the triangle at the scheduled or realtime location.

Alternatively, if they really are different: Since `VehicleAdherenceStatus` is just used for the `LadderVehicle` and vehicle icon, maybe `off-course` should be a view concept, and this definition could be moved out of this file (which is just for data), and into a file that works more with view stuff. Maybe `vehicleStatus`?

arkadyan replied:
Ok

arkadyan commented on assets/src/models/ladderVehicle.ts:
```diff
@@ -1,11 +1,27 @@
   y: number
   vehicleDirection: VehicleDirection
+  scheduledY?: number
+  scheduledVehicleDirection?: VehicleDirection
```
The whole point of these was to show intermediate stages of data and make it clear that certain functions only require some subsets of properties. `LadderVehicle` is the only thing that matters publicly though, so I think it's valuable to keep everything explicitly defined there, rather than having to try to decipher the results of a chain of `extends`.

skyqrose commented on assets/src/models/vehicleStatus.ts:
```diff
@@ -0,0 +1,28 @@
+    ? "off-course"
+    : vehicle.scheduleAdherenceStatus
+
+export const isUnassignedBySwiftly = ({
```
I think one reason I wanted to combine `VehicleAdherenceStatus` and `ScheduleAdherenceStatus` is so a bunch of files don't need to know details about Swiftly. Another way to accomplish the same goal could be to rename this function to (or export a different function) `isOffCourse` instead of `isUnassignedBySwiftly`, so the callers can think about the meaning rather than the implementation.

arkadyan replied:
I like that 👍
